src/c_api/c_api.cc

```diff
@@ -758,6 +758,39 @@ XGB_DLL int XGDMatrixGetUIntInfo(const DMatrixHandle handle,
   API_END();
 }
 
+XGB_DLL int XGDMatrixCopyDataCSR(const DMatrixHandle handle,
+                                 size_t **out_row_ptr,
+                                 unsigned **out_indices,
+                                 float **out_data) {
+  API_BEGIN();
+  CHECK_HANDLE();
+  const auto dmat = static_cast<std::shared_ptr<DMatrix>*>(handle)->get();
+
+  *out_row_ptr = reinterpret_cast<size_t *>(
```

Comment:
You need to specify who owns the memory.  If XGBoost owns the memory then we also need to create a `free` function.  On the other hand if the user owns this memory we should have a way to tell how should they allocate memory, like size of `values` and `row_ptr_`.  I personally prefer the second option.

Reply:
Thanks for detailed comments. Since these method will be created language bindings other than C/C++, maybe the first approach is better? After read all you comments, I think this PR need to be refactored into a class like `DMatrixView`?  In this way, memory can be freed  
 by destructor.